web/src/app/services/user.service.ts

```diff
@@ -3,42 +3,81 @@ import { from, Observable } from 'rxjs';
 import { map, catchError } from 'rxjs/operators';
 import {
   AngularFirestore,
-  AngularFirestoreCollection
+  AngularFirestoreCollection,
+  AngularFirestoreDocument
 } from 'angularfire2/firestore';
 import { ErrorService } from './error.service';
+import { AuthService } from './auth.service';
+import { User } from '../models/user.model';
+import { UserSocial } from '../models/userSocial.model';
 
 @Injectable()
 export class UserService {
   public user: AngularFirestoreCollection;
   public userSocial: AngularFirestoreCollection;
 
   constructor(
-    db: AngularFirestore,
-    private errorService: ErrorService ) {
-    this.userSocial = db.collection('userSocial');
-    this.user = db.collection('user');
+    private db: AngularFirestore,
+    private errorService: ErrorService,
+    private authService: AuthService,
+  ) {
+    this.userSocial = this.db.collection('userSocial');
+    this.user = this.db.collection('user');
   }
-  saveUser(user) {
-    return from(this.user.add({ user }))
+
+  saveUser(user: User): Observable<User> {
+    return from(this.user.doc(user.uid).set({ user }))
       .pipe(
-        map((response) => this.addRefID(response, user)),
+        map(() => this.addRefID(user)),
         catchError((err) => this.errorService.logError(err)),
       );
   }
 
-  addRefID(response, user) {
-    const normalisedResponse = { ...user.additionalUserInfo.profile, reference: response.id };
+  getUser(): Observable<User> {
+    return this.authService.user.pipe(
+      map(response => this.formatResponse(response))
+    );
+  }
+
+  getUserSocialDetails(providerId: string, uid: string): Observable<UserSocial> {
+    return from(this.userSocial.ref.where(`${providerId}.reference`, '==', uid).get())
+      .pipe(
+        map(response => this.getSocialDataFromPayload(response)),
+        catchError(error => this.errorService.logError(error))
+      );
+  }
+
+  getSocialDataFromPayload(response): UserSocial | null {
+    let social;
+    if (!response) {
+      return null;
+    }
+    response.docs.forEach((doc) => {
+      social = doc.data();
```

Comment:
This will only take the latest result from `doc.data()`

Reply:
@eddiejaoude Couldn't get you. Could you please elaborate?

Reply:
Because this is in a loop, `social` will get overwritten each time and only the last result of the loop will be kept in `social`.


For example
```typescript
const a = [1,2,3];

let b = 0;

a.forEach((x) => {
   b = x;
}

console.log(b); // will always equal the last value of the array, in this case 3
```

Reply:
Got it @eddiejaoude . Thanks for the code snippet 😄 . As we query only for one social doc from the firebase, the array will always have one element, so I extracted the data of that element from the array and used social variable to store that particular detail. So didn't push or use an array here.

Reply:
Ok, if there is only ever a single item in the array, using a `forEach` seems a bit overkill. What about just getting the first element with `a[0]` or `a.pop()`?

Reply:
sure will change and commit